README.md
# WebFPGA - FPGA IceStorm Examples

This repository is set up for command-line development or online development using [Gitpod](https://gitpod.io). The `src` directory has a couple of examples to get you started:

* blinky.v: blinks the LED once a second using the internal oscillator
* top.v: fades the on-board RGB Neopixel

## If you are using Gitpod (online editing and compiling)

1. [Fork](https://docs.github.com/en/get-started/quickstart/fork-a-repo) this repo
2. Prefix the URL with `gitpod.io/#` like this: `gitpod.io/#https://github.com/yourname/webfpga-icestorm-example`
3. Enjoy editing in Visual Studio Code or JetBrains

Use `make` to build your code. 

**Note** You can't flash from Gitpod yet. We're working on it.

## If you are working from the command line

1. Clone this repo
2. Install the [OSS CAD Suite](https://github.com/YosysHQ/oss-cad-suite-build) which includes Yosys, NextPNR, Verilator, etc.

Use `make` to build your code. Use `make flash proj=` and your project name to flash the FPGA.

## Change from the example on Youtube

* The makefile has been completely redone. You no longer need to edit the makefile to specify a project. Instead, add it to the `make flash` command like this: `make flash proj=blinky`.
* Sample code moves to the `src` directory. `make all` builds all of it.
* Tooling moves to the `tools` directory. 

CHANGELOG.md
# WebFPGA Icestorm Examples

## [1.0.0] Example as shipped with WebFPGA

## [1.1.0] 

### Using make

* The makefile has been completely redone. 
	* You no longer need to edit the makefile to specify a project. 
	* Add the project to `make flash`  like this: `make flash proj=blinky`.
	* Bitstream compression is inline in the make file, removing the need for `bin_to_bc.sh`.
	* `make all` builds everything in `src`

### Pinmaps

* `src/pinmap.pcf` contains the hard-wired perhipherals on the Shasta Plus board. 

### Layout changes
* Sample code moves to the `src` directory.
* Code builds into the `build` directory.
* Tooling moves to the `tools` directory. 

## [1.1.1]

Added configuration file for [gitpod.io](https://gitpod.io). This gives you a browser-based editor 
along all with all the build tools. (But no USB-based flashing, yet.)

## [1.1.2]

Implements `@FPGA_TOP`.

## [1.1.3]

Turns on linting via verilator.

rgb.v
module fpga_top (
    output reg  WF_NEO
);

 
// define registers
//
// 1 pixel
// these are register, they can be moved to a RAM for larger pixel
// arrays. Also since they are 8 bits each, likely more compact
// storage could be used by reducing the color from 8 bits to
// 6 or 5 bits, without much loss of colors.

   reg  [23:0] pixel0;

   reg  [4:0]  pixel_bit_cnt;
   reg  [1:0]  pixel_cnt;
   reg  [23:0] shift_pixels;

   wire        phase;
   reg         phase_d;
   reg         phase_tmp;

   reg         np_eof;
   wire        np_sof;

   reg [2:0] reset_cnt;
   reg [3:0] np_clk_cnt;
   reg       pixel_bit_end;
   reg       tog_bit0;
   reg       tog_bit1;

//////////////////////////////////////////////////////////////////////
//
//  create clock from SB_HFOCS hard macro
  wire clk_en;
  assign clk_en = 1'b1;

  SB_HFOSC OSC_i (                  // 48 MHz clock
        .CLKHFEN(clk_en),
        .CLKHFPU(clk_en),
        .CLKHF(clk));

    // set to 12MHz (ie divide by 4)
   defparam OSC_i.CLKHF_DIV = "0b10";

///////////////////////////////////////////////////////////////////////
//  send data stream continuously {50 us gap, then pixel0}.

   always @(posedge clk)
//     begin
//     if (rst)
//       begin
//       end
//     else
         begin 
           if (phase==0) // neopixel reset
             begin
            if (np_sof)
           begin
                   WF_NEO      <= 1'b1;  // every bit starts at logic 1
              end
               else
                 begin
                   WF_NEO         <= 1'b0;
                   shift_pixels <= pixel0;
                   pixel_bit_cnt<= 5'b0;  // 24 bits per pixel
                   np_eof       <= 1'b0;  // clr singal to create pulse
                 end
             end
           else
             begin
               if (( shift_pixels[0] && tog_bit1) ||
                   (~shift_pixels[0] && tog_bit0) )
                  WF_NEO <= 1'b0;

//       at end of bit time, shift the pixels or load new 24 bits
               if (pixel_bit_end)
                 if (pixel_bit_cnt != 23)
                   begin
                     WF_NEO          <= 1'b1;
                     shift_pixels  <= {1'b0,shift_pixels[23:1]};
                     pixel_bit_cnt <= pixel_bit_cnt + 1'b1;
                   end
           else
             begin     // start reset cycle
                     np_eof <= 1'b1;
                     WF_NEO   <= 1'b0; 
             end
             end
         end


/////////////////////////////////////////////////////////////////////////
// create pulses for bit0 and bit1 switching to low state.
// sync on np_sof signal.
//
// one bit stream is:
//  
//    |                             |
//    |-------------                -----------
//    |             |               |
//    |<- hi time ->|< lo time    ->|
//    |             |               |
//    |              ----------------
//    |                             |
//    |<-         bit time        ->|
//
//    first bit out is the msb for color intentsity.
//    GREEN followed by RED followed by BLUE.
//
//  table values are in ns and +/- tolerances
//
//            period   |   1 hi  |   0 hi  |   1 lo  |   0 lo  |  reset 
//=========================================================================
// WS2812B  1250/600   | 800/150 | 400/150 | 450/150 | 850/150 | >50 us


// this device has 12 MHz, so 83.333ns period. Table for clocks and time

//            period   |   1 hi  |   0 hi  |   1 lo  |   0 lo  |  reset 
//=========================================================================
// WS2812B  15/1250    | 10/833  |  5/417  |  5/417  | 10/833  | >600 clks
//
//
parameter [3:0] BIT1_HI = 10;
parameter [3:0] BIT1_LO =  5;
parameter [3:0] BIT0_HI =  5;
parameter [3:0] BIT0_LO = 10;

parameter [3:0] NEO_PERIOD = 15;


// create bit waveform
   always @(posedge clk)
     begin
       if (np_sof)
         begin
           np_clk_cnt <= 4'b1;  // advance 1 cnt due to dout starting early
                         // for first transfer.
         end
       else if (phase == 1'b1)
         begin
           if (np_clk_cnt == NEO_PERIOD- 4'h1)
             begin
               np_clk_cnt     <= 4'b0;
               pixel_bit_end  <= 1'b1;
             end
           else
             np_clk_cnt <= np_clk_cnt + 4'b1;

           if (np_clk_cnt == BIT0_HI -4'h1)
             tog_bit0 <= 1'b1;     // pulse to change dout for bit being 0
           else
             tog_bit0 <= 1'b0;

           if (np_clk_cnt == BIT1_HI -4'h1)   // pulse to change dout for bit being 1
             tog_bit1 <= 1'b1;
           else
             tog_bit1 <= 1'b0;

           if (pixel_bit_end) pixel_bit_end <= 1'b0; // create pulse
         end
     end

/////////////////////////////////////////////////////////////////////////
// setup approx timers
//
     wire ten_us;
     wire two_ms;

     // include time base
     ten_usec  ten_usec ( 
           .clk(clk),
           .ten_us(ten_us));

     // include time base
     two_msec two_msec( 
           .clk(clk),
           .ten_us(ten_us),
           .two_ms(two_ms));

/////////////////////////////////////////////////////////////////////////
// 50us reset counter  -pause after complete data stream of string of LEDs

   always @(posedge clk)
     begin
       if (phase_tmp == 0)
      begin
           if (ten_us)
             begin
               if (reset_cnt == 'h6)
                 begin
                   phase_tmp <= 1'b1;  // start data frame
                   reset_cnt <= 'd0; // clr rst cnt readying it for nxt use
                 end
               else
                 reset_cnt <= reset_cnt + 3'd1;
             end  
         end

       if (np_eof)           // once last data is sent start reset again
         phase_tmp <= 1'b0;

// create delays for SOF
       phase_d <= phase_tmp;

     end

    assign np_sof =  ~phase_d && phase_tmp; 
    assign phase = phase_d;
// 
//////////////////////////////////////////////////////////////////
//  Create colors to DEMO above bit logic.
//  Loading register or memory from SPI or of means is possible
//////////////////////////////////////////////////////////////////
// create colors in pxiel registers.

  reg [1:0] cnt;
  reg [6:0] cnt2;
  reg       updn;
  reg [2:0] ph;

  wire [6:0]  cnt2_G;
  wire [6:0]  cnt2_R;
  wire [6:0]  cnt2_B;

// bit reverse for WS2812B
  wire [6:0] cnt2_w={cnt2[0],cnt2[1],cnt2[2],cnt2[3],cnt2[4],cnt2[5],cnt2[6]};

  assign cnt2_G = cnt2_w;
  assign cnt2_R = cnt2_w;
  assign cnt2_B = cnt2_w;


  // up-down counters for fading
   always @(posedge clk)
     begin
       if (two_ms)
      begin
           cnt <= cnt + 2'h01;   // every XXms change the color
           if (cnt == 0)
             begin
               if (cnt2 ==  127)
              begin
                   cnt2 <= 126;
                   updn = ~updn;
                ph <= ph + 3'b1;
              end
            else if (cnt2 == 1 && updn ==1)  // down counting
                begin
                     cnt2 <= 1;
                     updn = ~updn;
                  ph <= ph + 3'b1;
                end
            else if (updn==0)
                     cnt2 <= cnt2 + 7'b1;
            else
                     cnt2 <= cnt2 - 7'b1;
        end   // cnt == 0
     end      // 10 ms
    end
        


   always @(*)
    case (ph)   // set up neopixel color cycling (all are half brightness)
      3'h0: pixel0[23:0] <= {8'h00,8'hfe,8'h00};    // red
      3'h1: pixel0[23:0] <= {8'h00,8'h00,8'hfe};    // green
      3'h2: pixel0[23:0] <= {8'hfe,8'h00,8'h00};    // blue
      3'h3: pixel0[23:0] <= {8'h00,{cnt2_R,1'b0},8'h00};  // fade red
      3'h4: pixel0[23:0] <= {8'h00,8'h00,{cnt2_G,1'b0}};  // fade green
      3'h5: pixel0[23:0] <= {{cnt2_B,1'b0},8'h00,8'h00};  // fade blue
      3'h6: pixel0[23:0] <= {8'h00,{cnt2_R,1'b0},{~cnt2_G,1'b0}}; // mix RG
      3'h7: pixel0[23:0] <= {{cnt2_B,1'b0},8'h00,{~cnt2_G,1'b0}}; // mix BG
    endcase


endmodule

//////////////////////////////////////////////////////////////////////
//
// 10us timer pulse
//
//
   module ten_usec (
       input  wire clk,
       output reg  ten_us
   );

    reg [6:0] counter;

    always @ (posedge clk)
        if (counter == 119)
          begin
            ten_us   <= 1'b1;
            counter  <= 'b0;
          end
        else
          begin
            counter  <= counter + 7'b1;
            ten_us   <= 1'b0;
          end

   endmodule

//////////////////////////////////////////////////////////////////////
//
// 2ms timer pulse
//
//
   module two_msec (
       input  wire clk,
       input  wire ten_us,
       output reg  two_ms
   );

    reg [9:0] counter;

    always @ (posedge clk)
        if (ten_us)
       begin
            if (counter == 199)
              begin
                two_ms   <= 1'b1;
                counter  <=   'b0;
              end
           else
            counter  <= counter + 10'b1;
          end
        else
            two_ms   <= 1'b0;

   endmodule


src/blinky.v
// blinky.v
//
// The "Hello World" of Digital Logic!
// This example blinks the on-board user LED once a second.
//
// This Source Code is Public Domain.

// @MAP_IO clk    13

module fpga_top(input wire clk, output reg WF_LED);
    reg [24:0] counter;

    // Blink LED every 1000 ms
    always @ (posedge clk) begin
        // Try using a different value here...
        // 12000000 will cause it to blink twice as fast!
        if (counter == 8000000) begin
            WF_LED   <= ~WF_LED;
            counter  <= 'b0;
        end else begin
            counter  <= counter + 'b1;
        end
    end

    // Setup CLK using internal oscillator
    //
    // We will add more documentation on this later. But basically, this
    // is a Lattice specific macro to use internal oscillator as the clock.
    // The WebFPGA board also has an external oscillator for precision
    // timing applications.
    //wire clk_en;
    //assign clk_en = 1'b1;
    //SB_HFOSC OSC_i(
        //.CLKHFEN(clk_en),
        //.CLKHFPU(clk_en),
        //.CLKHF(clk));
endmodule


tools/get_top.py
#!/usr/bin/env python3

# Copyright 2022 Google LLC2
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# author: Sarah Clark <sarahclark@google.com>

"""
Parses the @FPGA_TOP directive in a Verilog file and writes to standard output
"""
import sys
import re

top_comment = re.compile(r'(?i)\s*//\s*@FPGA_TOP\s+(\w+)')


def main():
    top_module = 'fpga_top'
    if len(sys.argv) == 2:
        fname = sys.argv[1];
        with open(fname) as source_file:
            for line in source_file:
                match_result = top_comment.match(line)
                if match_result is not None:
                    top_module = match_result.group(1)
                    break
    print(top_module)

if __name__ == '__main__':
    main()


src/WF_RGB_dot_matrix_demo.v
// @MAP_IO RGB_DOUT    25 
// @MAP_IO RGB_CLK_OUT 26 
// @MAP_IO RGB_LOAD    27 

// #CAS_CLK pin OSC_i/CLKHF 83.3    //ns   

module fpga_top(
    output wire RGB_CLK_OUT,
    output wire RGB_DOUT,
    output wire RGB_LOAD,
    input  wire WF_BUTTON,
    input  wire WF_CPU1,
    input  wire WF_CPU2,
    input  wire WF_CPU3
);


   // input IO from CPU
   wire      cpuin_pa0_toggled;
   reg       cpuin_pa0_d;

   // ram write to demo display driver
   reg         updates;   // allow updates to memory only when not scanning
   reg [15:0]  ram_wr_pixels;
   reg [5:0]   ram_wr_addr;
   reg         ram_wr_en;
   wire [15:0]  ram_rd_pixels;
   wire [5:0]   ram_rd_addr;
   wire [4:0]  random_num;
   wire        colR;
   wire        colG;
   wire        colB;
   reg  [2:0]  cnt;

/////////////////////////////////////////////////////////////////////////
//
//  create clock for SB_HFOCS hard macro
// OSC enable
   wire      clk_en = 1'b1;
    

  SB_HFOSC OSC_i (
        .CLKHFEN(clk_en),
        .CLKHFPU(clk_en),
        .CLKHF(clk));

    // set to 12MHz
   defparam OSC_i.CLKHF_DIV = "0b10";

     // include time base  10us
     WF_timer #(.COUNT(119), .EVERY_CLK(1'b1)) ten_usec( 
           .clk(clk),
           .enable(1'b1),
           .timer_pulse(ten_us));
 
     // include time base  2ms
     WF_timer #(.COUNT(199)) two_msec(   
           .clk(clk),
           .enable(ten_us),
           .timer_pulse(two_ms));

     // include time base  1 sec
     WF_timer #(.COUNT(499)) one_sec( 
           .clk(clk),
           .enable(two_ms),
           .timer_pulse(one_second));

//
//  debounce switch in

    WF_switch_debounce switch_deb1 (
      .clk(clk),              // main clock
      .sample_en(two_ms),     // pulse when to sample
      .switch_in(WF_BUTTON),  // async switch input
      .switch_out(),          // sync switch out, debounced
      .switch_pushed(switch_pushed),    // pulse when switch is pushed
      .switch_released()                // pulse when switch is released
     );


// detect CPU in toggling
//
   always @ (posedge clk)
       cpuin_pa0_d        <= WF_CPU1;

     assign cpuin_pa0_toggled = ~cpuin_pa0_d &&  WF_CPU1 ||
                                 cpuin_pa0_d && ~WF_CPU1;

//
// use memory to store 8x8 pixel matrix colors.

   WF_bram256x16 dot_memory (
     .clk(clk),
     .wen(ram_wr_en),
     .ren(1'b1),
     .waddr({2'b0,ram_wr_addr}),
     .raddr({2'b0,ram_rd_addr}),
     .wdata(ram_wr_pixels),
     .rdata(ram_rd_pixels));

// bring in the RGB I/F
    WF_RGB_if RGB_if_inst(
     .clk(clk),
     .scan_en(two_ms),
     .scan_done(scan_done),
     .ram_rd_addr(ram_rd_addr),
     .ram_rd_pixels(ram_rd_pixels),
     .CLK_OUT(RGB_CLK_OUT),
     .LOAD(RGB_LOAD),
     .DOUT(RGB_DOUT)
     );

///////////////////////////////////////////////////////////////////

     // **************************************
     //  fill memory to test display.
     //
    
 // sync to the scan done sugnals to updates to memory
 // when scanning isn't being done.
 //
     always @ (posedge clk)
       if (two_ms)
         updates <= 1'b0;
       else if (scan_done)
           updates <= 1'b1;
         else if (~WF_CPU2)   // if doing random hen only
           updates <= 1'b0;    // update memory once every two ms

     always @ (posedge clk)
       if (updates)   // update when not scanning
        begin
            ram_wr_addr <= ram_wr_addr + 1;
            ram_wr_pixels <= {1'b0,4'b0,colR, 4'b0,colG,4'b0, colB}; 
//            ram_wr_pixels <= {1'b0,4'b0,1'b0, 4'b0,1'b1,4'b0,1'b0}; 
            ram_wr_en <= 1'b1;
        end
        else
            ram_wr_en <= 1'b0;

     always @ (posedge clk)
       if (switch_pushed || cpuin_pa0_toggled)
         cnt <= cnt + 1;


    // get random 5 bit number.
     WF_random_5 random5 (
           .clk(clk),
           .enable(two_ms),
    	   .data(random_num));

    assign colR = WF_CPU2 ? cnt[0] : random_num[0];
    assign colG = WF_CPU2 ? cnt[1] : random_num[1];
    assign colB = WF_CPU2 ? cnt[2] : random_num[2];

////////////////////////////////////////////////////////////////////////////

endmodule

////////////////////////////////////////////////////////////////////////////
//
// RGB I/F
// 
//////////////////////////////////////////////////////////////////////
//  RGB dot matrix display driver and logic
//
//  This code is written for a serial 8x8 3 color dot matrix display.
//  It is wired as follows:
//
//  8 row drivers enabled when high.  (ANODE)
//  3 - 8 column sinks, LED on when low. (CATHODE)
//  There are 3 row sinks, one for each color (red,green,blue).
//
//  Shift 32 bits of data per scan.
//      {row,colG,colR,colB}, each are 8 bits
//
//      MSB bit is shift first.
//  Need to generate a CLK, DOUT and LOAD signal.
//  DOUT is changed on falling edge of CLK. LOAD is pulsed high
//  after last CLK falling edge.
//////////////////////////////////////////////////////////////////////
//
  module WF_RGB_if(
	  input  clk,
          input  scan_en,   // start scan cycle, two_ms min
	  output scan_done, // can be used to sync updates

	  output [5:0]  ram_rd_addr,   // pixel memory addr
	  input  [15:0] ram_rd_pixels, // pixel data {1'b0,5'RED,5'GRN,5'BLUE}
	                           // currently no PWM, so LSB of each color 
	  output reg  CLK_OUT,  // send to IOs
	  output wire LOAD,
	  output wire DOUT
  );

  wire scan_done = dotclk_done;

// variables
  reg        phase_col;
  reg        phase_rd;
  reg [5:0]  dotclk_index;
  reg [2:0]  dotclk_index_d;
  reg        dotclk_en;      // during DOUT shifting
  reg        dotclk_en_d;    // delayed
  wire       dotclk_done;    // after falling edge of dotclk_en
  reg        dotclk_done_d; 
  reg        phase_row;      // on when row selects are being shifted out
  reg [7:0]  row;            // scan enables 
  reg [2:0]  row_addr;
  reg [23:0] data_next;      // these can be optimized, use only 8 bits
  reg [23:0] data_out; 

// scan rows every scan_en, so compete display is updated every 8 scan enables,
// 16ms, is the minimum scan o no see the flickering.

// row scanning 
     always @ (posedge clk)
       begin
         if (row == 8'h0)     // just came out of reset
           row <= 8'hfe;

         if (dotclk_done )
           row <= {row[6:0],row[7]};   // rotate the select bit
       end

////////////////////////////////////////////////////////////////////////////
//  Main logic controller. Starts when scan_en pulse is true.
//  read memory phase, 8 clocks to read pxiel RAM, one row, or 8 bytes.
//  start shifting data, 24 bits,
//  finally shift out row selects 8 bits.
//  end cycle
//

    always@ (posedge clk)   
      if (scan_en)      // start cycle
   //   if (WF_CPU3, ? one_second:scan_en)      // start cycle
        phase_rd <= 1'b1;
      else
        begin         
          if (dotclk_index == 15 && phase_rd)  // end of phase_rd
            begin
              phase_rd     <= 1'b0;
              phase_col    <= 1'b1;
              dotclk_index <= 0;
              dotclk_en    <= 1'b1;   // enable external clock
            end
          else 
	    if (phase_rd || phase_col || phase_row)
	      dotclk_index <= dotclk_index + 'b1;

          if (dotclk_index == 47)     // end of phase_col
            begin
              phase_col    <= 1'b0;
              phase_row    <= 1'b1;  // enable loading for row data
            end
          if (dotclk_index == 6'd63) // check for last bit
            begin
              dotclk_en    <= 1'b0;
              phase_row    <= 1'b0;  
            end
        end      

////////////////////////////////////////////////////////////////////////////

// lookup pixel from memory
//
// take row and convert to address.
   always @(*)                                                                
         case (row)  //synthesis parallel_case full_case 
             8'b1111_1110 : row_addr <= 3'h0;
             8'b1111_1101 : row_addr <= 3'h1;
             8'b1111_1011 : row_addr <= 3'h2;
             8'b1111_0111 : row_addr <= 3'h3;
             8'b1110_1111 : row_addr <= 3'h4;
             8'b1101_1111 : row_addr <= 3'h5;
             8'b1011_1111 : row_addr <= 3'h6;
             8'b0111_1111 : row_addr <= 3'h7;
         endcase                                                                

// get next row pixel data from memory
//  use pixel index to help construct ram addr.
    assign ram_rd_addr =  {row_addr,dotclk_index[2:0]};

    // below supports only 8 colors, need to add PWM to mix the colors
    always@ (posedge clk)   
      if (phase_row) // reading data while row sel is shifting
         begin
           dotclk_index_d <= dotclk_index[2:0];  // ram pipeline
     
           data_next[dotclk_index_d[2:0]]    <= ram_rd_pixels[0];   // blue
           data_next[dotclk_index_d[2:0]+16] <= ram_rd_pixels[5];   // green
           data_next[dotclk_index_d[2:0]+8]  <= ram_rd_pixels[10];  // red
       end


////////////////////////////////////////////////////////////////////////////
//
      // generate clock
     always @ (posedge clk)     // clock will be half core clock, 6MHz.
       if (dotclk_en)
         CLK_OUT <= ~CLK_OUT;
       else
         CLK_OUT <= 1'b0;   // make sure it at zero at idle

// send data out 32 clocks of data, { colG,colR,colB, row}
//
//  use 64 clocks, data will change on every other, so data changes on falling
//  edge out the output-ed clock
//
     always @ (posedge clk)
       if (dotclk_en )
         begin
           if (dotclk_index == 47)    // load pixel data for the row
             data_out <= {row, 16'b0};
           else 
             if (dotclk_index[0]) // only change on falling edge
             data_out <= {data_out[22:0],1'b0};  // shift data out
         end
       else            // load next data continueously until dotclk_en
         data_out <= data_next;

     always @ (posedge clk)
       begin
         dotclk_en_d   <= dotclk_en;   // delay so read data is ready
         dotclk_done_d <= dotclk_done; // delay to load new row after scan
       end   

     // send signals out
     assign LOAD = ~dotclk_en;
     assign DOUT = data_out[23];

     assign dotclk_done = ~dotclk_en && dotclk_en_d;  // create pulse

endmodule

//////////////////////////////////////////////////////////////////////
//
// timer 
//
   module WF_timer (
       input  wire clk,
       input  wire enable,
       output reg  timer_pulse
   );

   parameter [9:0] COUNT = 499;
   parameter       EVERY_CLK = 1'b0;

    reg [9:0] counter;

    always @ (posedge clk)
        if (enable)
          begin
            if (counter == COUNT)
              begin
                timer_pulse <= 1'b1;
                counter     <= 'b0;
              end
           else
             begin
              counter     <= counter + 'b1;
              if (EVERY_CLK)  // case enable never goes low
               timer_pulse <= 1'b0;
             end

          end
        else
            timer_pulse <= 1'b0;

   endmodule

//////////////////////////////////////////////////////////////////////
//
// pseudo random sequence
// 5 bit output.
// pattern repeats every 32 counts
//
   module WF_random_5 (
     input  wire      clk,
     input  wire      enable,
     output reg [4:0] data
   );

   wire [4:0] data_next;
  
   // unrolled equations
   assign data_next[4] = data[4]^data[1];
   assign data_next[3] = data[3]^data[0];
   assign data_next[2] = data[2]^data_next[4];
   assign data_next[1] = data[1]^data_next[3];
   assign data_next[0] = data[0]^data_next[2];

   always @(posedge clk)
     if (enable)
       if(data == 0)  // just out of reset, or all zeros case
         data <= 5'h1f;
       else
         data <= data_next;

   endmodule

//////////////////////////////////////////////////////////////////////
//
// 256x16 FPGA RAM, BRAM.
//
   module WF_bram256x16 (
     input wire         clk,
     input wire         wen,
     input wire         ren, 
     input wire  [7:0]  waddr, 
     input wire  [7:0]  raddr,
     input wire  [15:0] wdata,
//     output reg  [15:0] rdata
     output wire [15:0] rdata
     );

     // inferred BRAM, doesn't work, so use manual instantiation
     reg [15:0] mem [0:63];

     always @(posedge clk)
       begin
         if (wen)
           mem[waddr] <= wdata;

         if (ren)
           rdata <= mem[raddr];
       end

/*
 SB_RAM256x16  SB_RAM256x16_inst (  
                      .WDATA(wdata),
                      .MASK(16'b0),
                      .WADDR(waddr),
                      .WE(wen),
                      .WCLKE(1'b1),
                      .WCLK(clk),

                      .RDATA(rdata),
                      .RADDR(raddr),                             
                      .RE(1'b1),
                      .RCLKE(1'b1),
                      .RCLK(clk));
*/
   endmodule

//////////////////////////////////////////////////////////////////////
//
// external switch debounced and synchronizer.
// active low when pushed.
//
   module WF_switch_debounce (
     input  wire clk,              // main clock
     input  wire sample_en,        // pulse when to sample
     input  wire switch_in,        // async switch input
     output reg  switch_out,       // sync switch out, debounced
     output wire switch_pushed,    // pulse when switch is pushed
     output wire switch_released); // pulse when switch is released

     reg  [2:0] switch;
     reg        switch_out_d;
//
     always @ (posedge clk)
       begin
         if ( sample_en)
           switch <= { switch[1:0],switch_in };

         if (switch == 3'b000)
           switch_out <= 1'b1;
         else if (switch == 3'b111)
           switch_out <= 1'b0;
     
         switch_out_d <= switch_out;
       end

     assign switch_pushed   = ~switch_out &&  switch_out_d;
     assign switch_released =  switch_out && ~switch_out_d;

   endmodule


tools/map_io.py
#!/usr/bin/env python3

# Copyright 2022 Google LLC
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# author: Sarah Clark <sarahclark@google.com>

"""
Parses the @MAP_IO directive in a Verilog file and writes a Physical Constraints file (.pcf).
"""
import argparse
import logging
import re

map_comment = re.compile(r'(?i)\s*//\s*@MAP_IO\s+(\w+)(\[\d+\])?\s+(\d+)')

PCF_HEADER = """
set_io -nowarn WF_LED       31
set_io -nowarn WF_CLK       35
set_io -nowarn WF_BUTTON    42
set_io -nowarn WF_NEO       32
set_io -nowarn WF_CPU1      11
set_io -nowarn WF_CPU2      12
set_io -nowarn WF_CPU3      13
set_io -nowarn WF_CPU4      10
"""


def main():
    """Parse the command line and process the files"""
    parser = argparse.ArgumentParser(allow_abbrev=False)

    group = parser.add_mutually_exclusive_group()
    group.add_argument("-v", "--verbose", help="display all files processed",
                       action="store_const", const=logging.INFO, dest='loglevel')
    group.add_argument("-d", "--debug", help="display debugging information",
                       action="store_const", const=logging.DEBUG, dest='loglevel')

    parser.add_argument('-o', '--output', default='pinmap.pcf',
                        help='output file with .pcf extension')
    parser.add_argument("-n", "--dry-run", help="examine files without changing",
                        action="store_true")
    parser.add_argument("file", nargs=1, help="file to check")
    args = parser.parse_args()
    logging.basicConfig(level=args.loglevel or logging.WARNING)

    if args.dry_run:
        logging.info("Dry run -- no files will be overwritten")

    symtable = []
    for fname in args.file:
        with open(fname) as source_file:
            logging.info(r"in: %s", fname)
            symtable = [mapping for line in source_file if (
                mapping := get_mapping(line)) is not None]

    if not symtable:
        logging.warning('No @MAP_IO statements found for %s', args.output)

    if not args.dry_run:
        logging.info(r'out: %s', args.output)
        with open(args.output, mode='wt') as pcf_file:
            pcf_file.write(PCF_HEADER)
            for item in symtable:
                pcf_file.write(item.to_set_io())


def get_mapping(line):
    """ Extract the mapping for this line. Returns None if not a map directive"""
    match_result = map_comment.match(line)
    if match_result is None:
        return None
    key = match_result.group(1)
    index = match_result.group(2)
    value = int(match_result.group(3))
    if index is not None:
        index = int(index.removeprefix('[').removesuffix(']'))
    return MapSpec(key, value, index)


class MapSpec:
    """Holds the values from the @MAP_IO statement"""

    def __init__(self, key, value, index=None) -> None:
        # Map external pin 1 to internal pin 17, etc.
        pin_map = [17, 16, 14, 23, 20, 19, 18, 21, 25, 26, 28, 27, 34, 35,
                   36, 37, 40, 44, 46, 47, 45, 48, 2, 3, 4, 9, 6, 43, 41, 39, 38, 15]
        self.symbol = f"key${index}" if index is not None else key
        self.base_name = key
        self.ext_pin = value
        self.int_pin = pin_map[value]
        self.index = index

    def __hash__(self):
        return hash((self.symbol))

    def to_set_io(self) -> str:
        return f"set_io\t{self.symbol}\t{self.int_pin}\n"


if __name__ == '__main__':
    main()
